.pre-commit-config.yaml

```diff
@@ -0,0 +1,63 @@
+ci:
+  autoupdate_commit_msg: "chore: Update pre-commit hooks"
+exclude: ^(\.data/|\.scratch/)
+repos:
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: "v0.4.1"
+    hooks:
+      - id: ruff
```

Comment:
In what relationship is ruff to the other tools? Which tools will be replaced (black, autopep, mypy etc) and which will stay?

Reply:
I suggest removing isort and flake8 as those tools are almost fully covered. Black can arguably also be removed without too many concerns, but as black is fairly fast, it doesn't hurt too much (they are compatible). Pylint does some things differently, I'd keep it for now. On a greenfield project, I'd rather not use it though.

----

https://docs.astral.sh/ruff/faq/#how-does-ruffs-formatter-compare-to-black

### Is the Ruff linter compatible with Black?

Yes. The Ruff linter is compatible with [Black](https://github.com/psf/black) out-of-the-box, as long as the [line-length](https://docs.astral.sh/ruff/settings/#line-length) setting is consistent between the two.

### How does Ruff's linter compare to Pylint?

At time of writing, Pylint implements ~409 total rules, while Ruff implements over 800, of which at least 209 overlap with the Pylint rule set (see: [#970](https://github.com/astral-sh/ruff/issues/970)).

### How does Ruff's linter compare to Flake8?

Ruff can be used as a drop-in replacement for Flake8 when used (1) without or with a small number of plugins, (2) alongside Black, and (3) on Python 3 code.

Ruff can also replace Black, isort, yesqa, eradicate, and most of the rules implemented in pyupgrade.

### How does Ruff's formatter compare to Black?
The Ruff formatter is designed to be a drop-in replacement for [Black](https://github.com/psf/black).

Specifically, the formatter is intended to emit near-identical output when run over Black-formatted code

### How does Ruff compare to Mypy, or Pyright, or Pyre?
Ruff is a linter, not a type checker.

Reply:
Sounds good. Let's remove them in a follow up PR where we enable ruff.